File: stack/non_coprime_replacements/non_coprime_replacements.cpp
#include <vector>
#include <numeric>
using namespace std;

vector<int> replace_non_coprimes(vector<int> &nums)
{ // LeetCode Q.2197.
    vector<int> non_corime_nums;
    for (auto num : nums)
    {
        while (!non_corime_nums.empty())
        {
            int greatest_common_divisor = gcd(num, non_corime_nums.back());
            if (greatest_common_divisor == 1)
            {
                break;
            }
            num *= non_corime_nums.back() / greatest_common_divisor;
            non_corime_nums.pop_back();
        }
        non_corime_nums.push_back(num);
    }
    return non_corime_nums;
}

File: stack/2197_hard/non_coprime_replacements.py
import math


def replace_non_coprimes(nums: list[int]) -> list[int]:  # LeetCode Q.2197.
    replacements = [nums.pop(0)]
    for num in nums:
        gcd = math.gcd(replacements[-1], num)
        while replacements and gcd > 1:  # Keep replacing previous replacements.
            num *= replacements.pop(-1) // gcd
            if replacements:
                gcd = math.gcd(replacements[-1], num)

        replacements.append(num)

    return replacements
